### PR TITLE
Some function pointers, typedefs, and OSX's stdlib.

### PR DESCRIPTION
**Pull request by emilio**
r? @nox 

The change from indexing cursors to index USRs was because in the `inner_template_self` test, the `next` pointer pointed to a clang-generated class declaration that caused us to not catch the type as already resolved.

Also, that's the recommended way to check against same symbols from different translation units, though now we don't use it yet.

### CODE REVIEW

File: tests/expectations/blocks.rs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+/* automatically generated by rust-bindgen */
+
+
+#![allow(non_snake_case)]
+
+
+extern "C" {
+    pub fn atexit_b(arg1: ::std::option::Option<unsafe extern "C" fn()>);
```
Comment (nox) [inline tests/expectations/blocks.rs]: We should just bind blocks to `*mut c_void` given their ABI is weird.

File: src/ir/context.rs [hunk 2]
```diff
@@ -131,29 +141,46 @@ impl<'ctx> BindgenContext<'ctx> {
...
-                      (location.unwrap().kind() == CXCursor_ClassTemplate ||
-                       location.unwrap().kind() == CXCursor_ClassTemplatePartialSpecialization) {
-                let old = self.types.insert(location.unwrap().canonical(), id);
-                debug_assert_eq!(old, None);
-            } else {
+            let mut declaration = declaration.unwrap();
+            if !declaration.is_valid() {
+                if let Some(location) = location {
+                    if location.kind() == CXCursor_ClassTemplate ||
+                       location.kind() == CXCursor_ClassTemplatePartialSpecialization {
```
Comment (emilio) [inline src/ir/context.rs]: Yeah, I want to use the `matches!` crate instead, here and in a few other places, but I'll do that in a follow-up.

File: src/ir/context.rs [hunk 3]
```diff
@@ -131,29 +141,46 @@ impl<'ctx> BindgenContext<'ctx> {
...
+                        declaration = location;
+                    }
+                }
+            }
+            declaration = declaration.canonical();
+            if !declaration.is_valid() {
```
Comment (nox) [inline src/ir/context.rs]: Here you rebound `declaration` to the canonical one, which is invalid.

What does the `debug!` message end up saying, given the declaration you passed to it is the invalid one now?

**Pull request merged**